library/alloc/src/vec/mod.rs

```diff
@@ -2792,6 +2792,51 @@ impl<T, I: SliceIndex<[T]>, A: Allocator> IndexMut<I> for Vec<T, A> {
     }
 }
 
+/// Collects an iterator into a Vec, commonly via [`Iterator::collect()`]
+///
+/// # Allocation behavior
+///
+/// In general `Vec` does not guarantee any particular grow/allocation stategy.
+/// That also applies to this trait impl.
+///
+/// **Note:** This section covers implementation details and is therefore exempt from
+/// stability guarantees.
+///
+/// Vec may use any or none of the following strategies,
+/// depending on the supplied iterator:
+///
+/// * preallocate based on [`Iterator::size_hint()`]
+///   * and panic if the number of items is not outside the provided lower/upper bounds
+/// * use an amortized growth strategy similar to `pushing` one item at a time
+/// * perform the iteration in-place on the original allocation backing the iterator
+///
+/// The last case warrants some attention. It is an optimization that in many cases reduces peak memory
+/// consumption and improves cache locality. But when a large number of big, short-lived
+/// allocations are created, only a small fraction of their items gets collected, no further use
+/// is made of the spare capacity and the resulting `Vec` is moved into a longer-lived structure
+/// this can lead to the large allocations having their lifetimes unnecessarily extended which
+/// can result in increased memory footprint.
```

Comment:
This a long run-on sentence -- maybe:

```suggestion
/// allocations are created, and only a small fraction of their items get collected, no further use
/// is made of the spare capacity. If the resulting `Vec` is moved into a longer-lived structure,
/// this can lead to the large allocations having their lifetimes unnecessarily extended which
/// can result in increased memory footprint.
```

But this scenario about "a large number" of allocations seems a bit _too_ specific to me. It could just as well be one giant vector that gets reduced without any reclaim. I think the point we want to get across is that we may reuse your buffer even if the result has much more capacity that it needs.

Reply:
If you only do it once then you already paid the max-RSS cost with the temporary allocation. You have to do something like this more than once to blow up max-RSS.

Reply:
Focusing on RSS assumes an OS that overcommits memory, which isn't universal. And even in that case, all it takes is for the code to fill the "original" vector again to pump up RSS.

Reply:
If you build a giant vector, and then filter-collect it to a much smaller size, you might reasonably expect that giant memory block to now be available for other allocations. This can have an impact even just once.

Reply:
> This a long run-on sentence -- maybe

It's supposed to be an "But when a), b) and then c) [happen] this can result X."
The suggested `.` broke it to "But when a), b). If c) [happens] this can result X."

Reply:
Hmm, yes that's broken -- that's what I get for not quoting the whole thing when trying to revise it.